### PR TITLE
Add inline annotations and avoid 128-bit where it's not needed in TryFrom + tests for xsize/x128

### LINKED ISSUE #43124: Collect<Vec<u16>> from range doesn't optimize well.
(At least on x86-64 nightly)

Using code like this: 
https://is.gd/nkoecB

The version using collect is significantly slower than creating a vec of 0-values and setting the values manually.

```
test using_collect ... bench:     117,777 ns/iter (+/- 6,424)
test using_manual  ... bench:       7,677 ns/iter (+/- 365)
test using_unsafe  ... bench:       3,866 ns/iter (+/- 394)
```

On the other hand, if using u32 instead with the same code collect is much better:
```
test using_collect ... bench:       7,677 ns/iter (+/- 555)
test using_manual  ... bench:      12,487 ns/iter (+/- 836)
test using_unsafe  ... bench:       7,741 ns/iter (+/- 413)
```

Same with u64: 
```
test using_collect ... bench:      18,675 ns/iter (+/- 1,335)
test using_manual  ... bench:      29,692 ns/iter (+/- 1,864)
test using_unsafe  ... bench:      18,559 ns/iter (+/- 1,065)
```

I suspect this may be SIMD-related. Will see if there are similar results on stable.

### PR DESCRIPTION
**Pull request by oyvindln**
This should make the output in #43064 slightly cleaner, and hopefully help fix #43124 when combined with #43127. I think the main issue was the lack of inline annotations for `try_from` (and not the use of 128-bit values as I first thought). Additionally, the implementations of try_from called min/max_value through the FromStrRadixHelper trait, which weren't marked inline either. I would have thought LLVM would be smart enough to inline this anyway, but it seems it isn't. While there has been talk about [moving to a `Cast`]( #42456) trait for number types, it looks like it will take a while before that can happen, so this should help a bit in the meantime.

I didn't do any pointer width specific implementations for now. 

This change also makes the assumption that i/usize is not larger than 64 bits wide to simplify things a little. Is that a safe assumption to make for now?

All the tests passed, but it's still possible I could have missed some parts.

This PR:
Avoids using i/u128 except for converting between 128-bit types.
   Use the smallest type able to represent the maximum value instead for most types (except u/isize), and i/u64 otherwise.

Add sinline annotations to try_from implementations.
   Before doing this the functions did not get inlined.

Avoids using the FromStrRadixHelper trait in the try_from implementations, as the trait wasn't actually needed for anything.

Add some simpler implementations for conversions that don't need as many checks. In release mode the checks will be optimized out when they're not needed anyhow, but this should at least result in very slightly less code non-optimized builds and very slightly less work for LLVM.

### CODE REVIEW

**Comment by aidanhs** [comment]
Thanks for the PR @oyvindln, we'll check in now and again to make sure @aturon or another reviewer gets to this soon!

**Review by nagisa — CHANGES_REQUESTED** [review]

File: src/libcore/num/mod.rs [hunk 1]
```diff
@@ -2523,42 +2524,91 @@ macro_rules! same_sign_try_from_int_impl {
...
+trivial_try_from_impl!(u128, u128);
+trivial_try_from_impl!(usize, usize, u128, i128);
+
+trivial_try_from_impl!(i8, i8, i16, i32, i64, i128, isize);
+trivial_try_from_impl!(i16, i16, i32, i64, i128);
+trivial_try_from_impl!(i32, i32, i64, i128);
+trivial_try_from_impl!(i64, i64, i128);
+trivial_try_from_impl!(i128, i128);
+trivial_try_from_impl!(isize, isize, i128);
+trivial_try_from_impl!(usize, u64);
```
Comment (oyvindln) [inline src/libcore/num/mod.rs]: Ah, simply an oversight, you're right it should be together with the other unsigned ones.

File: src/libcore/num/mod.rs [hunk 2]
```diff
@@ -2523,42 +2524,91 @@ macro_rules! same_sign_try_from_int_impl {
...
+            #[inline]
+            fn try_from(u: $source) -> Result<$target, TryFromIntError> {
+                // Using into as an extra check on the macro arguments.
+                Ok(u as $target)
+            }
+        }
+    )*}
+}
 
+// The current implementations assume that usize/isize <= 64 bits wide.
```
Comment (oyvindln) [inline src/libcore/num/mod.rs]: Makes sense.

File: src/libcore/num/mod.rs [hunk 3]
```diff
@@ -2523,42 +2524,91 @@ macro_rules! same_sign_try_from_int_impl {
...
+same_sign_try_from_wider_impl!(u16, u32, u64, u128);
+same_sign_try_from_wider_impl!(i16, i32, i64, i128);
+same_sign_try_from_wider_impl!(u32, u64, u128);
+same_sign_try_from_wider_impl!(i32, i64, i128);
+same_sign_try_from_wider_impl!(u64, u128);
+same_sign_try_from_wider_impl!(i64, i128);
+
+// Need to make sure the storage component is large enough for these.
+// (storage, target, $(source))
+same_sign_try_from_int_impl!(u64, usize, u64, u32, u16);
```
Comment (oyvindln) [inline src/libcore/num/mod.rs]: I initially decided to leave that for later as it seemed a bit complex to fully test, but I suppose I can try to do it properly.

**Comment by oyvindln** [comment]
Thanks for the feedback. I will see if I can sort out the issues pointed out.

**Comment by oyvindln** [comment]
Okay, I think I sorted the requested changes.

I'm a bit unsure about the "compile_error" macro call after checking for an unknown bit width. I didn't enable the compile_error feature, as that would require me to add it to lib.rs, and add a statement to ignore the warning about unused feature. Though, at least it won't compile if the condition is hit, even though it won't give the proper compile_error message.

I didn't manage to compile for the 16-bit target (I got an missing specification error for the msp430_none_elf target which I don't know how to solve.), so that part is untested still.

I also couldn't find any tests for the usize and u128 types for TryFrom. Maybe that's something that ought to be added.

**Comment by SimonSapin** [comment]
> I didn't manage to compile for the 16-bit target (I got an missing specification error for the msp430_none_elf target which I don't know how to solve.), so that part is untested still.

(Perhaps https://github.com/rust-lang/rust/pull/43099 has not reached stage 0 yet? Before https://github.com/rust-lang-nursery/rust-forge/pull/80 the platform support page linked to https://github.com/pftbest/rust_on_msp/blob/master/msp430.json.)

**Comment by oyvindln** [comment]
Yup, that seems to be the issue. I'm trying to build with rustc set to the latest nightly (which does have msp430 support) in config.toml now.

**Comment by nagisa** [comment]
#[cfg_attr(any(target_poin...), feature(compile_error))] should work for the conditional feature.

**Comment by nagisa** [comment]
@oyvindln do you want to add the conversion tests for usize & u128? If not, I’ll review it as is.

**Comment by oyvindln** [comment]
@nagisa You can review it as is for now.

File: src/libcore/num/mod.rs [hunk 4]
```diff
@@ -2523,42 +2524,120 @@ macro_rules! same_sign_try_from_int_impl {
...
+same_sign_try_from_wider_impl!(u8, u16, u32, u64, u128, usize);
+same_sign_try_from_wider_impl!(i8, i16, i32, i64, i128, isize);
+same_sign_try_from_wider_impl!(u16, u32, u64, u128);
+same_sign_try_from_wider_impl!(i16, i32, i64, i128);
+same_sign_try_from_wider_impl!(u32, u64, u128);
+same_sign_try_from_wider_impl!(i32, i64, i128);
+same_sign_try_from_wider_impl!(u64, u128);
+same_sign_try_from_wider_impl!(i64, i128);
+same_sign_try_from_wider_impl!(usize, u128);
+same_sign_try_from_wider_impl!(isize, u128);
```
Comment (nagisa) [inline src/libcore/num/mod.rs]: Shouldn’t this be `isize, i128`?

File: src/libcore/num/mod.rs [hunk 5]
```diff
@@ -2523,42 +2524,120 @@ macro_rules! same_sign_try_from_int_impl {
...
+    }
+
+    // Same for 16 and 32-bit platforms.
+    #[cfg(any(target_pointer_width = "16", target_pointer_width = "32"))] {
+        // xsize -> x32
+        trivial_try_from_impl!(usize, u32);
+        trivial_try_from_impl!(isize, i32);
+        // x64 -> xsize
+        same_sign_try_from_wider_impl!(usize, u64);
+        same_sign_try_from_wider_impl!(isize, u64);
```
Comment (nagisa) [inline src/libcore/num/mod.rs]: A cross-sign conversion again.

I think this code would be easier follow if it was only split into blocks for each target_pointer_width.

Something like:

```
#[cfg(target_pointer_width = "16")] { 
     trivial_try_from_impl!(usize, u16, u32);
     trivial_try_from_impl!(isize, i16, i32);
     same_sign_try_from_wider_impl!(usize, u32, u64);
     same_sign_try_from_wider_impl!(isize, i32, i64);
}

// and so forth.
```

Comment (nagisa) [inline src/libcore/num/mod.rs]: Also, I believe the argument order is swapped for `same_sign_try_from_wider_impl` here and in the block above?

File: src/libcore/num/mod.rs (hunk 4 above)
Comment (oyvindln) [inline src/libcore/num/mod.rs]: Indeed

File: src/libcore/num/mod.rs (hunk 5 above)
Comment (oyvindln) [inline src/libcore/num/mod.rs]: It make make it more readable yeah.

**Comment by oyvindln** [comment]
I borrowed the idea of using from/into as a separator in the macros from the try_from crate. Hopefully that makes it a bit easier to follow.

Also there is a new PR  #43248 that does something similar to this PR. The approach used there may be a bit cleaner.

**Comment by nagisa** [comment]
I’m personally fine with either approach (they both look fairly similar to me). @llogiq needs to at least add tests and the `#[inline]` annotations -- something your PR has and their doesn’t.

If you feel that their implementation is better, I would suggest @llogiq to simply take your tests and merge that.

/me shrugs

**Comment by oyvindln** [comment]
If you think this one is fine, feel free to review/merge.

**Comment by llogiq** [comment]
@oyvindln are you OK with me cherry-picking your PR?

**Comment by oyvindln** [comment]
@llogiq Yup, that's fine by me.

**Comment by alexcrichton** [comment]
ping r? @nagisa

**Comment by carols10cents** [comment]
ping @nagisa @rust-lang/libs, it seems like this is ready for rereview!

**Comment by nagisa** [comment]
See [this](https://github.com/rust-lang/rust/pull/43248#issuecomment-317484644).

**Comment by oyvindln** [comment]
I think #43248 will be used instead. (Though I can fix up this to add the modules and avoid storage types if you don't feel like waiting for the other PR to be finished.)

**Comment by oyvindln** [comment]
Looks like #43248 is ready now, so I'm closing this in favour of that.

**Pull request closed without merging**